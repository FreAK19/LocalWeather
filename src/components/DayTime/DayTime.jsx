import React, {Component} from 'react';
import PropTypes from 'prop-types'
import './DayTime.less'

export default class DayTime extends Component {
  dayOfWeek = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  monthOfYear = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ];
  state = {
    now: new Date()
  };

  static propTypes = { caption: PropTypes.string.isRequired };

  componentDidMount() {
    this.timer = setInterval(() => {
      this._tick();
    }, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.timer);
  }

  _tick() {
    this.setState({
      now: new Date()
    })
  }

  render() {
    const {now} = this.state;
    return (
      <div className='info'>
        {
          this.props.error ?
            <h3 className='info__caption'> Please turn on Geolocation on Browser </h3> :
            <h3 className='info__caption'>{this.props.caption}</h3>
        }
        <p>{this.dayOfWeek[now.getDay()]}</p>
        <p>
          {this.monthOfYear[now.getMonth()]} {now.getDate()}, {now.getFullYear()}
        </p>
        <span>
          {
            now.getHours() + ':' + now.getMinutes()
          }
          {now.getHours() > 12 ? ' PM' : ' AM'}</span>
      </div>
    )
  }
}