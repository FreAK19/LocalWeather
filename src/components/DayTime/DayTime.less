
.info {
  text-align: left;
  flex-basis: 210px;
  padding-right: 1rem;

  &__caption {
    font-size: 2.7rem;
    font-weight: 900;
    line-height: 1;
    margin-top: 0;

    &.small {
      font-size: 1.7rem;
    }
  }

  & p {
    font-size: 1.4rem;
    font-weight: 700;
    margin: 0;
  }
  & p:nth-child(3) {
    font-weight: 400;
    font-size: 1.1rem;
  }
}