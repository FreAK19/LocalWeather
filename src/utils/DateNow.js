export default class DateNow extends Date {
  constructor() {
    super();
    this.now = new Date();
    this.daysOfWeek = [
      'Sunday',
      'Monday',
      'Tuesday',
      'Wednesday',
      'Thursday',
      'Friday',
      'Saturday'
    ];
    this.monthsOfYear = [
      'January',
      'February',
      'March',
      'April',
      'May',
      'June',
      'July',
      'August',
      'September',
      'October',
      'November',
      'December'
    ];
  }

  getDate() {
    return this.now.getDate();
  }

  getHours() {
    return this.now.getHours();
  }

  getFullYear() {
    return this.now.getFullYear();
  }

  getHoursToString() {
    return this.now.toString().match(/\d+:\d+/)[0];
  }

  getDayToString() {
    return this.daysOfWeek[this.now.getDate()];
  }

  getMonthToString() {
    return this.monthsOfYear[this.now.getMonth()];
  }
}