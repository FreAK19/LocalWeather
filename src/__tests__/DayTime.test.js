import React from 'react';
import DayTime from '../components/DayTime/DayTime';
import DateNow from '../utils/DateNow';

describe('DayTime', () => {
  test('should render correctly', () => {
    const date = new DateNow('02/05/2018 18:39');
    const component = shallow(<DayTime caption="Geolocation is not support" error={false} />);
    component.setState({ now: date });
    expect(component).toMatchSnapshot();
  });

  test('should render error message coerces to boolean true', () => {
    const component = shallow(<DayTime caption="Geolocation is not support" error="some error" />);
    const expected = 'Please turn on Geolocation on Browser';
    const actual = component.find('.info__caption').text();
    expect(actual).toMatch(expected);
  });

  test('should render PM if hour is over 12', () => {
    const component = shallow(<DayTime caption="Kiev" error={false} />);
    component.setState({ now: new DateNow('12/12/2015 12:00') });
    const text = component.find('span').text();
    const actual = /PM/g.test(text);
    expect(actual).toBe(true);
  });

  test('should render AM if hour is less then 12', () => {
    const component = shallow(<DayTime caption="Kiev" error={false} />);
    component.setState({ now: new DateNow('12/12/2015 6:00') });
    const text = component.find('span').text();
    const actual = /AM/g.test(text);
    expect(actual).toBe(true);
  });

  test('should be equal to expected object of Time', () => {
    const component = shallow(<DayTime caption="London" error={false} />);
    component.setState({ now: new DateNow('02/11/2018 16:20') });
    const expected = ['Saturday', '16:20', 'February'];
    const state = component.state();
    const actual = [state.now.getDayToString(), state.now.getHoursToString(), state.now.getMonthToString()];
    expect(actual).toMatchObject(expected);
  });
});
