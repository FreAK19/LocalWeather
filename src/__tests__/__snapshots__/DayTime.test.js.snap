// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DayTime should render correctly 1`] = `
<div
  className="info"
>
  <h3
    className="info__caption"
  >
    Geolocation is not support
  </h3>
  <p>
    Monday
  </p>
  <p>
    February
     
    5
    , 
    2018
  </p>
  <span>
    18:39
     PM
  </span>
</div>
`;
